src/pages/resultsView/plots/PlotsTab.tsx

```diff
@@ -0,0 +1,1110 @@
+                entrezGeneId: this.vertSelection.entrezGeneId!
+            }));
+        }
+        if (promises.length) {
+            return remoteData({
+                await: ()=>promises,
+                invoke:()=>Promise.resolve(_.flatten(promises.map(x=>x.result!)))
+            });
+        } else {
+            return undefined;
+        }
+    }
+
+    @computed get horzAxisDataPromise() {
+        return makeAxisDataPromise(
+            this.horzSelection,
+            this.clinicalAttributeIdToClinicalAttribute,
+            this.props.store.molecularProfileIdToMolecularProfile,
+            this.props.store.patientKeyToSamples,
+            this.props.store.entrezGeneIdToGene,
+            this.props.store.clinicalDataCache,
+            this.props.store.numericGeneMolecularDataCache,
+        );
+    }
+
+    @computed get vertAxisDataPromise() {
+        return makeAxisDataPromise(
+            this.vertSelection,
+            this.clinicalAttributeIdToClinicalAttribute,
+            this.props.store.molecularProfileIdToMolecularProfile,
+            this.props.store.patientKeyToSamples,
+            this.props.store.entrezGeneIdToGene,
+            this.props.store.clinicalDataCache,
+            this.props.store.numericGeneMolecularDataCache
+        );
+    }
+
+    readonly mutationDataExists = remoteData({
+        await: ()=>[this.props.store.studyToMutationMolecularProfile],
+        invoke: ()=>{
+            return Promise.resolve(!!_.values(this.props.store.studyToMutationMolecularProfile).length);
+        }
+    });
+
+    readonly cnaDataExists = remoteData({
+        await: ()=>[this.props.store.studyToMolecularProfileDiscrete],
+        invoke: ()=>{
+            return Promise.resolve(!!_.values(this.props.store.studyToMolecularProfileDiscrete).length);
+        }
+    });
+
+
+    @computed get horzLabel() {
+        if (this.props.store.molecularProfileIdToMolecularProfile.isComplete &&
+            this.props.store.entrezGeneIdToGene.isComplete) {
+            return getAxisLabel(
+                this.horzSelection,
+                this.props.store.molecularProfileIdToMolecularProfile.result,
+                this.props.store.entrezGeneIdToGene.result,
+                this.clinicalAttributeIdToClinicalAttribute
+            );
+        } else {
+            return "";
+        }
+    }
+
+    @computed get vertLabel() {
+        if (this.props.store.molecularProfileIdToMolecularProfile.isComplete &&
+            this.props.store.entrezGeneIdToGene.isComplete) {
+            return getAxisLabel(
+                this.vertSelection,
+                this.props.store.molecularProfileIdToMolecularProfile.result,
+                this.props.store.entrezGeneIdToGene.result,
+                this.clinicalAttributeIdToClinicalAttribute
+            );
+        } else {
+            return "";
+        }
+    }
+
+    @computed get horzDescription() {
```

Comment:
i'm sort of torn about this, but i don't think we should be checking status of promises inside computed getters.   it hides the status of the promises from consuming code.   we need to make sure that loading status is handled properly at the view level and this pattern will let devs off that hook.  that is, components do not render and invoke these "getters" until they know the data is loaded.   instead of this, i would advocate we use pure functions which accept data as arguments.  in order to get the optimization of @computed, we can use @memoize which caches the result of functions based on arguments.  i.e. will not recompute result if it's done so already for equal arguments.

Reply:
in this particular case i think it actually makes sense because the return type is just a string for a tooltip (and in `horzLabel` and `vertLabel` its the string for the axis labels), and we wouldn't be showing a loading spinner anyway

Reply:
k